File: haskell/sum-of-multiples/src/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples factors limit = sum multiples
  where multiples = filter isMultiple [1..limit-1]
        isMultiple number = any (number `divisibleBy`) factors
        number `divisibleBy` factor = factor > 0 && number `mod` factor == 0

File: c/sum-of-multiples/sum_of_multiples.c
#include "sum_of_multiples.h"

unsigned int sum(const unsigned int * const factors,
                 const size_t nfactors,
                 const unsigned int limit) {
  unsigned int result = 0;

  if (factors == NULL) {
    return result;
  }

  for (unsigned int i = 1; i < limit; i++) {
    for (size_t j = 0; j < nfactors; j++) {
      if (factors[j] > 0 && i % factors[j] == 0) {
        result += i;
        break;
      }
    }
  }

  return result;
}
